### PR TITLE
fix(mitigation): skip the coefficient write on a kl_control hold

### LINKED ISSUE #371: Reward-hack controller: pid_lagrangian has never had a valid run, the rollback ladder never fires, and a non-acting kl_control run is numerically identical to log_only
## Three facts about one shipped feature, from the H100 session

These are separate from #286 (which asks whether the controller *works*). These say that parts of it have never been exercised **mechanically**, which is a prerequisite for asking about efficacy at all.

### 1. `pid_lagrangian` has no valid arm at all

All three attempts crashed, at steps **6, 25 and 8** (`benchmarks/gate-h100-validation.md` L3834–3836). That is a shipped `--reward-hack-mitigation` mode with **zero** successful runs — unexercised mechanically as well as statistically. (The crashes are #342's `grad_norm nan`, so this is partly blocked on that, but the mode has never been shown to complete a run even once.)

### 2. The rollback ladder never fired in any arm

The escalation ladder requires **three** consecutive HACK votes; the beta ladder escalated on **four** consecutive HACK votes in the same run (0.03 → 0.045 → 0.0675 → 0.1013 → 0.1519, each commanded value read back off the live trainer). The record states plainly: "this record does not explain why one path reached its threshold and the other did not. **Recorded as an open inconsistency.**" (L3829–3833)

Either the two thresholds are counted against different signals, or one of the counters resets where the other does not. Both are bugs; which one it is has not been established.

### 3. A `kl_control` run that never acts is numerically identical to `log_only`

`_run_bang_bang` calls `_apply_coefficient` on **every** step, including a `hold`. With `grpo_beta` and `reward_hack_beta_floor` both at the default `0.02`, that write puts back the value already there (L4602–4607).

This matters because **10 of 11** `kl_control` arms in the replication were in exactly that state — so any measured difference between the two modes in that pool is "a difference between two runs the code says are the same computation" (L4600–4612). It makes the attrition asymmetry (Fisher p=0.0275 unbalanced, p=0.0995 balanced) uninterpretable rather than merely unresolved.

## Fix path

1. Give `pid_lagrangian` one completed run at a configuration that does not trip #342 (lower lr, shorter horizon) before any efficacy claim is attempted.
2. Reconcile the two thresholds: make the rollback ladder and the beta ladder count the same signal with the same reset rule, or document why they differ.
3. Skip the coefficient write on a `hold` so a non-acting run is observably a no-op, and log the distinction, so "acted" vs "never acted" is recoverable from the mitigation log rather than inferred.

## Acceptance criteria

- [ ] One completed `pid_lagrangian` run exists, with its mitigation log
- [ ] A test that N consecutive HACK votes fires *both* ladders at their documented thresholds, with a control at N-1
- [ ] `hold` performs no write, and the mitigation log distinguishes acted / held / released
- [ ] The `beta == floor` degenerate case has a test — it is the configuration the defaults produce

### PR DESCRIPTION
**Pull request by AmirF194**
## Root cause

`_run_bang_bang` (the `kl_control` mode of the reward-hack mitigation controller) calls
`_apply_coefficient(action.new_beta)` on every training step, including a `hold`. On a
`hold`, `bang_bang_step` leaves `reason == "hold"` exactly when `new_beta == beta` (the
dead-band case, and the two clamped-at-ceiling/floor cases), so the write puts back the
value that was already there. With the default `grpo_beta` and `reward_hack_beta_floor`
both at `0.02`, most `kl_control` runs spend most of their steps in `hold`, so a run that
never trips the controller is, step for step, a same-value write repeated rather than the
no-op `log_only` claims to be.

## Fix

Skip `_apply_coefficient` when `action.reason == "hold"`. Since `hold` only occurs when
`new_beta == beta`, this changes no committed β; it only removes the redundant write. The
mitigation log now also carries `mitigation_status` (`held` / `acted` / `released`), so a
step where the controller did nothing is recoverable directly from the log instead of
inferred by parsing the free-text `action` field.

## Scope

This addresses fix-path item 3 of #371 only ("skip the coefficient write on a hold ... log
the distinction"), which is the `kl_control` sub-issue and is fully CPU-testable. Items 1
(`pid_lagrangian` needs a completed run on GPU hardware) and 2 (reconciling the rollback and
beta ladder thresholds, which needs the same GPU runs to observe) are unaddressed here; this
runner has no GPU access to produce or verify either. `Refs #371` rather than `Fixes #371`
for that reason.

## Verification

- `pytest tests/test_v07126.py -k TestKlControlCallback -v`: the three new tests fail against
  unpatched `main` (write-count assertions, missing `mitigation_status` key) and pass on this
  branch, checked in a clean Docker container on both ends of the CI matrix (Python 3.10 and
  3.12).
- Full `pytest tests/test_v07126.py` (190 tests, the whole reward-hack-mitigation suite):
  green on this branch, both Python versions.
- `ruff check src/soup_cli/ tests/`: clean. `mypy`: one pre-existing error in this file at
  line 922, unrelated to this change and present on `main` too (this repo's `type-check` CI
  job is non-blocking).
- Not run: anything requiring GPU (items 1/2 above), and the repository's full test suite
  (17k+ tests across 360 files) beyond the one file this change touches.

### CODE REVIEW

**Review by MakazhanAlpamys — APPROVED** [review]
Verified the load-bearing claim independently rather than taking it from the description, because the whole change rests on it.

**`reason == "hold"` really does imply `new_beta == beta`.** I walked all six paths through `bang_bang_step`: dead-band, trip-band-but-dwell-not-met, trip-band-already-tripped-clamped-at-ceiling, release-band-not-tripped, release-band-patience-not-met, and release-band-clamped-at-floor. In every one, `beta` leaves equal to what came in. So skipping the write commits the same β, and the mutation confirms the guard is watched:

```
$ # `if not held:` removed — write on every step
FAILED ...::TestKlControlCallback::test_hold_step_performs_no_coefficient_write
FAILED ...::TestKlControlCallback::test_hold_then_raise_writes_exactly_once
2 failed, 186 passed
```

`tests/test_v07126.py` 188 passed / 2 skipped otherwise.

## One edge, which I chased down and it is fine — but its safety is untested

There is a case where `hold` does **not** imply `new_beta == state.beta`:

```python
beta = state.beta if state.beta > 0.0 else policy.beta_floor
```

With the uninitialised sentinel `state.beta == 0.0`, a hold returns `new_beta == beta_floor`. Demonstrated:

```
state.beta   = 0.0
action.reason = 'hold'
action.new_beta = 0.02      <- differs from state.beta
```

Before this PR that value was written to the trainer; after it, it is not.

**It is unreachable in production, and I checked why**: `_run_bang_bang` calls `self._seed_coefficient(policy.beta_floor)` on the line *before* `bang_bang_step`, which resolves the sentinel from the live trainer coefficient. By the time the step runs, `state.beta > 0`. Your claim holds.

But that safety now depends on **two lines staying in that order, in a different method, with nothing asserting it**. Reorder them — or add an early return above the seed — and the first hold of a run silently stops establishing β, on a path where the symptom is a coefficient that is quietly never set rather than an exception.

One test would pin it:

```python
def test_the_sentinel_is_resolved_before_the_first_step():
    # state.beta == 0.0 must never reach bang_bang_step: on a hold it would
    # yield new_beta == beta_floor != state.beta, and the skip would then
    # change a committed value.
```

Assert on the *state* the step observes rather than on call order, so it survives a refactor that keeps the behaviour.

## The `mitigation_status` field is the better half of this PR

Recovering "the controller did nothing this step" by parsing free-text `action` was the kind of thing that works until someone rewords a message. `held` / `acted` / `released` as a first-class field is what makes #371's real question — *does `kl_control` differ from `log_only` when it never trips?* — answerable from the log instead of inferable from it.

---

Approving with the one test as an ask. Push it and I merge; if you would rather I add it on merge, say so.

**Comment by AmirF194** [comment]
Pushed: rebased onto main (the only conflict was two adjacent CHANGELOG entries, straightforward to merge) and added the regression test you asked for, pinning that state.beta is seeded from the live trainer coefficient before the first hold step rather than left at the sentinel.

Ran the full reward_hack_control suite locally (no torch in this environment, so the 9 tests that need it were skipped by import error, everything else green) plus ruff, both clean.

**Comment by MakazhanAlpamys** [comment]
Merged as [`64d4093`](https://github.com/MakazhanAlpamys/Soup/commit/64d4093) — your fourth. Entry is in [`CONTRIBUTORS.md`](https://github.com/MakazhanAlpamys/Soup/blob/main/CONTRIBUTORS.md).

**The ordering test does its job, and my first attempt at checking it was too weak — worth saying because it is the same trap I keep flagging in reviews.** I moved the seed below `_compute_vote` and nothing failed, which for a moment looked like the test was vacuous. It was my mutation that was wrong: `_compute_vote` does not touch `state.beta`, so the seed still landed before the step. Moving it below `bang_bang_step` — the reorder that actually lets the sentinel through — fails two:

```
FAILED ...::TestKlControlCallback::test_sentinel_is_resolved_before_the_first_hold_step
FAILED ...::TestKlControlCallback::test_ppo_kl_coef_mutated
2 failed, 187 passed
```

A mutation that does not fail proves nothing about the test until you have checked it was capable of failing.

Four merges, and what they have in common is worth naming: each one closed a gap between what a thing *reported* and what it *did*. `0/0` read as a pass, a pre-flight that could not see a per-process cap, a calibration hook with no caller, and now a controller whose log could not distinguish "held" from "acted". That is a specific taste in bugs and it is the useful kind.

#371 is not closed by this — the pid_lagrangian half and the rollback ladder are still unvalidated — but the part your PR names is now true: a non-acting `kl_control` run is a genuine no-op, and the log says so in a field.

**Comment by AmirF194** [comment]
Thanks, and good catch on your own mutation check, that is a sharp habit to have. Glad the acted/held/released distinction lands cleanly; happy to pick up the pid_lagrangian or ladder-reconciliation half later if either becomes buildable without GPU access.
